src/main/scala/chisel/lib/fifo/BubbleFifo.scala

```diff
@@ -0,0 +1,46 @@
+/*
+ * FIFO queue variations
+ *
+ * Author: Martin Schoeberl (martin@jopdesign.com)
+ */
+
+package chisel.lib.fifo
+
+import chisel3._
+
+/**
+  * A simple bubble FIFO.
+  * Maximum throughput is one word every two clock cycles.
+  */
+class BubbleFifo[T <: Data](gen: T, depth: Int) extends Fifo(gen: T, depth: Int) {
+
+  private class Buffer[T <: Data](gen: T) extends Module {
+    val io = IO(new FifoIO(gen))
+
+    val fullReg = RegInit(false.B)
+    val dataReg = Reg(gen)
+
+    when (fullReg) {
+      when (io.deq.ready) {
+        fullReg := false.B
+      }
+    } .otherwise {
+      when (io.enq.valid) {
+        fullReg := true.B
+        dataReg := io.enq.bits
+      }
+    }
+
+    io.enq.ready := !fullReg
+    io.deq.valid := fullReg
+    io.deq.bits := dataReg
+  }
+
+  private val buffers = Array.fill(depth) { Module(new Buffer(gen)) }
```

Comment:
I think many of these Modules may have problems in the degenerate case where the depth is zero. You might consider adding an assert or require to protect against this, or rewrite the loop to handle this.
The following more complex formulation handles the zero and other cases I think
```scala
  private val buffers = Array.fill(depth) { Module(new Buffer) }

  (Seq(io.enq) ++  buffers.map(_.io.deq)).zip(buffers.map(_.io.enq) :+ io.deq).foreach { case (prev, next) =>
    next <> prev
  }
```
It seems to me there is a constant tension in libraries like this between writing in a very plain style vs demonstrating the power of comprehensions. I'll leave that decision here for you.